test/program_test.go

```diff
@@ -0,0 +1,181 @@
+						},
+					},
+				},
+			},
+		}
+		prog.Name = randString()
+		prog.Namespace = "default"
+		Expect(k8sClient.Create(context.TODO(), &prog)).NotTo(Succeed())
+	})
+
+	It("should be possible to add a Program to a Stack", func() {
+		prog := pulumiv1.Program{}
+		prog.Name = randString()
+		prog.Namespace = "default"
+		Expect(k8sClient.Create(context.TODO(), &prog)).To(Succeed())
+
+		stack := pulumiv1.Stack{
+			Spec: shared.StackSpec{
+				ProgramRef: &shared.ProgramReference{
+					Name: prog.Name,
+				},
+			},
+		}
+		stack.Name = randString()
+		stack.Namespace = "default"
+
+		Expect(k8sClient.Create(context.TODO(), &stack)).To(Succeed())
+	})
+
+	When("Using a stack", func() {
+
+		var tmpDir, backendDir string
+		var kubeconfig string
+		var stack pulumiv1.Stack
+
+		BeforeEach(func() {
+			var err error
+			tmpDir, err = os.MkdirTemp("", "pulumi-test")
+			Expect(err).ToNot(HaveOccurred())
+
+			kubeconfig = writeKubeconfig(tmpDir)
+
+			backendDir = filepath.Join(tmpDir, "state")
+			Expect(os.Mkdir(backendDir, 0777)).To(Succeed())
+
+			stack = pulumiv1.Stack{
+				Spec: shared.StackSpec{
+					Stack:   randString(),
+					Backend: fmt.Sprintf("file://%s", backendDir),
+					EnvRefs: map[string]shared.ResourceRef{
+						"PULUMI_CONFIG_PASSPHRASE": shared.NewLiteralResourceRef("password"),
+						"KUBECONFIG":               shared.NewLiteralResourceRef(kubeconfig),
+					},
+				},
+			}
+			stack.Name = randString()
+			stack.Namespace = "default"
+		})
+
+		AfterEach(func() {
+			if strings.HasPrefix(tmpDir, os.TempDir()) {
+				os.RemoveAll(tmpDir)
+			}
+			err := k8sClient.Delete(context.TODO(), &stack)
+			if err != nil {
+				Fail(fmt.Sprintf("couldn't delete stack: %v", err))
+			}
+		})
+
+		It("should fail if given a non-existent program.", func() {
+
+			stack.Spec.ProgramRef = &shared.ProgramReference{
+				Name: randString(),
+			}
+
+			Expect(k8sClient.Create(context.TODO(), &stack)).To(Succeed())
+
+			waitForStackFailure(&stack)
+
+			Expect(stack.Status.LastUpdate.State).To(Equal(shared.FailedStackStateMessage))
+			stalledCondition := apimeta.FindStatusCondition(stack.Status.Conditions, pulumiv1.StalledCondition)
+			Expect(stalledCondition).ToNot(BeNil(), "stalled condition is present")
+			Expect(stalledCondition.Reason).To(Equal(pulumiv1.StalledSourceUnavailableReason))
+			Expect(apimeta.IsStatusConditionTrue(stack.Status.Conditions, pulumiv1.ReadyCondition)).To(BeFalse())
+		})
+
+		It("should fail if given a syntactically correct but invalid program.", func() {
+			prog := programFromFile("./testdata/test-program-invalid.yaml")
+			Expect(k8sClient.Create(context.TODO(), &prog)).To(Succeed())
+
+			stack.Spec.ProgramRef = &shared.ProgramReference{
+				Name: prog.Name,
+			}
+
+			Expect(k8sClient.Create(context.TODO(), &stack)).To(Succeed())
+
+			waitForStackFailure(&stack)
+
+			Expect(stack.Status.LastUpdate.State).To(Equal(shared.FailedStackStateMessage))
+			// TODO: Condition for invalid program?
```

Comment:
Generally all we'll know is that the automation API failed to run it to completion; a `failed` state might be as good as we can do for now.